nvflare/private/fed/app/simulator/simulator_runner.py

```diff
@@ -0,0 +1,383 @@
+        # Deploy the FL clients
+        self.logger.info("Create the simulate clients.")
+        for client_name in self.client_names:
+            client, self.client_config, self.deploy_args = self.deployer.create_fl_client(client_name, self.args)
+            self.federated_clients.append(client)
+            app_root = os.path.join(self.simulator_root, "app_" + client_name)
+            app_custom_folder = os.path.join(app_root, "custom")
+            sys.path.append(app_custom_folder)
+
+        self.logger.info("Set the client status ready.")
+        self._set_client_status()
+
+    def _set_client_status(self):
+        for client in self.federated_clients:
+            app_client_root = os.path.join(self.simulator_root, "app_" + client.client_name)
+            client.app_client_root = app_client_root
+            client.args = self.args
+            # self.create_client_runner(client)
+            client.simulate_running = False
+            client.status = ClientStatus.STARTED
+
+    def run(self):
+        try:
+            self.create_clients()
+            self.logger.info("Start the clients run simulation.")
+            executor = ThreadPoolExecutor(max_workers=self.args.threads)
+            lock = threading.Lock()
+            for i in range(self.args.threads):
+                executor.submit(lambda p: self.run_client_thread(*p), [self.args.threads, lock])
+
+            # wait for the server and client running thread to finish.
+            executor.shutdown()
+        except BaseException as error:
+            self.logger.error(error)
+        finally:
+            for client in self.federated_clients:
+                # client.engine.shutdown()
+                client.close()
+            # self.deployer.close()
+
+    def run_client_thread(self, num_of_threads, lock):
+        stop_run = False
+        interval = 1
+        client_to_run = None  # indicates the next client to run
+
+        try:
+            while not stop_run:
+                time.sleep(interval)
+                with lock:
+                    if not client_to_run:
+                        client = self.get_next_run_client()
+                    else:
+                        client = client_to_run
+
+                client.simulate_running = True
+                stop_run, client_to_run = self.do_one_task(client, num_of_threads, lock)
+
+                client.simulate_running = False
+        except BaseException as error:
+            self.logger.error(error)
+
+    def do_one_task(self, client, num_of_threads, lock):
+        open_port = get_open_ports(1)[0]
+        command = (
+            sys.executable
+            + " -m nvflare.private.fed.app.simulator.simulator_worker -o "
+            + self.args.workspace
+            + " --client "
+            + client.client_name
+            + " --port "
+            + str(open_port)
+        )
+        _ = subprocess.Popen(shlex.split(command, True), preexec_fn=os.setsid, env=os.environ.copy())
```

Comment:
should not need this

Reply:
This has to be in a separate child process. Pytorch / Ignite could not release thee GPU memory fully. Otherwise, we will  see GPU OOM over a period of time.